### PR TITLE
Add Yandex and Adfox ad networks integration for amp-ad.

### PR DESCRIPTION
**Pull request by DiMurer**

### CODE REVIEW

File: ads/adfox.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+function loadAdFox(global, cb) {
+  loadScript(global, 'https://yastatic.net/pcode/adfox/loader.js', cb);
+}
+
+/**
+ * @param {!Window} global
+ * @param {string} name
+ */
+function createContainer(global, name) {
+  const d = global.document.createElement('div');
```
Comment (DiMurer) [inline ads/adfox.js]: jfyi
When I tried to use the method `createElementWithAttributes` from `src/dom`, travis did not pass the test with error `Found forbidden: "win.open(" in dist.3p/current/integration.js:9897:11`.
It seems that this is an incorrect behavior.

**Comment by DiMurer** [comment]
Are there any problems with this pull request?
It expects a review for more than six days.

**Comment by DiMurer** [comment]
Still waiting for any feedback 😕

File: ads/adfox.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import {yandex} from './yandex';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function adfox(global, data) {
+
+  validateData(data, [
+    'adfoxParams',
```
Comment (zhouyx) [inline ads/adfox.js]: please inline if not exceed 80 chars limit.

File: ads/adfox.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  ]);
+
+  loadAdFox(global, () => initAdFox(global, data));
+}
+
+/**
+ * @param {!Window} global
+ * @param {!Function} cb
+ */
+function loadAdFox(global, cb) {
```
Comment (zhouyx) [inline ads/adfox.js]: Do we really need an extra function here? I'll say let's move `loadScript()` to `adfox()`.

File: ads/adfox.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  global.document.getElementById('c').appendChild(d);
+}
+
+/**
+ * @param {!Window} global
+ * @param {Object} data
+ */
+function initAdFox(global, data) {
+  const containerName = 'adfox_container';
```
Comment (zhouyx) [inline ads/adfox.js]: can we inline the `adfox_container` to containerId below?

File: ads/adfox.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        data-owner-id="208087"
+        data-adfox-params='{"pt": "b","p1": "bsoji","p2": "feil","pct": "a","pfc": "bbhfo","pfb": "cwrtv"}'>
+</amp-ad>
+```
+
+## Configuration
+
+For semantics of configuration, please see [documentation](https://specs.adfox.ru/page/254/).
+
+Supported parameters:
```
Comment (zhouyx) [inline ads/adfox.md]: As far as I see these two params are required. If so please indicate them as requried parameters.

File: ads/adfox.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+limitations under the License.
+-->
+
+# AdFox
+
+## Example
+
+```html
+<amp-ad width="360" height="400"
+        type="adfox"
```
Comment (zhouyx) [inline ads/adfox.md]: please fix indentation.

File: ads/yandex.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/yandex.js]: 2017

File: ads/yandex.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function yandex(global, data) {
+  validateData(data, ['blockId'], ['data', 'isAdfox']);
+
+  addToQueue(global, data);
+  loadScript(global, 'https://yastatic.net/partner-code/loaders/context_amp.js');
```
Comment (zhouyx) [inline ads/yandex.js]: please follow 80 chars limit.
Please also follow indentation rule when split to two line.

File: ads/yandex.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+ */
+function addToQueue(global, data) {
+  global[n] = global[n] || [];
+  global[n].push(() => {
+
+    // Create container
+    createContainer(global, renderTo);
+
+    // Show Ad in container
+    global.Ya.Context.AdvManager.render({
```
Comment (zhouyx) [inline ads/yandex.js]: is `global.Ya.Context.AdvManager` always there?

File: ads/yandex.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+limitations under the License.
+-->
+
+# Yandex
+
+## Example
+
+```html
+<amp-ad width="300" height="250"
+        type="yandex"
```
Comment (zhouyx) [inline ads/yandex.md]: please fix indentation.

File: ads/yandex.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+## Configuration
+
+For semantics of configuration, please see ad network documentation.
+
+
+Supported parameters:
+
+- data-block-id
```
Comment (zhouyx) [inline ads/yandex.md]: data-block-id is required parameter?
also please also document the other two supported parameters

File: examples/adfox.amp.html [hunk 12]
```diff
@@ -0,0 +1,71 @@
+<!doctype html>
```
Comment (zhouyx) [inline examples/adfox.amp.html]: We don't need example page for every ad network. Please only add the example to ads.amp.html page.

File: examples/ads.amp.html [hunk 13]
```diff
@@ -59,7 +59,7 @@
     <a href="#accesstrade">AccessTrade</a> |
     <a href="#adblade">Adblade</a> |
     <a href="#adbutler">AdButler</a> |
-    <a href="#adition">ADITION</a> |
+    <a href="#adfox">AdFox</a> |
```
Comment (zhouyx) [inline examples/ads.amp.html]: Please fix here.

File: examples/ads.amp.html [hunk 14]
```diff
@@ -252,6 +252,16 @@ <h2 id="adhese">Adhese</h2>
     <div fallback></div>
   </amp-ad>
 
+  <h2 id="adfox">AdFox</h2>
+
+  <amp-ad width="248" height="408"
+          type="adfox"
+          data-owner-id="208087"
+          data-adfox-params='{"pt": "b","p1": "bsoji","p2": "feil","pct": "a","pfc": "bbhfo","pfb": "cwrtv"}'>
+    <div placeholder></div>
```
Comment (zhouyx) [inline examples/ads.amp.html]: We now support default placeholder and fallback. No need to add placeholder and fallback here

File: examples/ads.amp.html (hunk 13 above)
Comment (zhouyx) [inline examples/ads.amp.html]: Please also add example to yandex ad.

File: ads/yandex.js (hunk 9 above)
Comment (DiMurer) [inline ads/yandex.js]: ` global.Ya.Context.AdvManager` is created after the `https://yastatic.net/partner-code/loaders/context_amp.js` is loaded.

**Comment by DiMurer** [comment]
Done

File: ads/yandex.js (hunk 9 above)
Comment (zhouyx) [inline ads/yandex.js]: But `addToQueue` is called before `loadScript`?

File: ads/adfox.md [hunk 15]
```diff
@@ -0,0 +1,35 @@
+<!---
```
Comment (zhouyx) [inline ads/adfox.md]: Need to link certain ad type readme from `amp-ad.md`

File: ads/adfox.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+function checkLoading(global, data) {
+  if (data.bundleName === 'banner.direct') {
+    const dblParams = {
+      blockId: data.bundleParams.blockId,
+      data: data.bundleParams.data,
+      isAdfox: true,
+    };
+
+    yandex(global, dblParams);
+    return false;
```
Comment (zhouyx) [inline ads/adfox.js]: does the function need to `return true` otherwise?

File: ads/yandex.md [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+For semantics of configuration, please see ad network documentation.
+
+
+Required parameters:
+- data-block-id
+
+Supported parameters:
+- data
+- isAdfox
```
Comment (zhouyx) [inline ads/yandex.md]: hmmm `data-isAdfox` will be converted to `data.isadfox`  Do you mean `data-is-adfox` here?

File: ads/yandex.md [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+## Configuration
+
+For semantics of configuration, please see ad network documentation.
+
+
+Required parameters:
+- data-block-id
+
+Supported parameters:
+- data
```
Comment (zhouyx) [inline ads/yandex.md]: `data-data` please.

File: examples/ads.amp.html [hunk 19]
```diff
@@ -1165,6 +1174,12 @@ <h2 id="yahoojp">YahooJP YDN</h2>
       data-yadsid="79712_113431">
   </amp-ad>
 
+  <h2 id="yandex">Yandex</h2>
```
Comment (zhouyx) [inline examples/ads.amp.html]: I am not able to see a rendered ad here. Please double check.

Comment (zhouyx) [inline examples/ads.amp.html]: I am not able to see a rendered ad here. Please double check.

File: ads/adfox.md (hunk 15 above)
Comment (DiMurer) [inline ads/adfox.md]: fixed

File: examples/ads.amp.html (hunk 19 above)
Comment (DiMurer) [inline examples/ads.amp.html]: Hmmm. This is strange. It is works for me https://yadi.sk/i/6T1S7uQN3J6Srz

Comment (DiMurer) [inline examples/ads.amp.html]: This is strange.
It is works for me https://yadi.sk/i/6T1S7uQN3J6Srz

**Comment by DiMurer** [comment]
Done

File: ads/adfox.js (hunk 16 above)
Comment (zhouyx) [inline ads/adfox.js]: Sure it really depends on your side implementation. As long as not calling `if(!checkingLoading)` or so.

File: ads/yandex.js (hunk 9 above)
Comment (zhouyx) [inline ads/yandex.js]: got it

File: examples/ads.amp.html (hunk 19 above)
Comment (zhouyx) [inline examples/ads.amp.html]: got 
image
Please test the example with our local server.

**Comment by zhouyx** [comment]
@DiMurer The pr looks good. We should be able to proceed with it once I am able to see the yandex ad example.

**Comment by DiMurer** [comment]
@zhouyx I made changes in the code. Now everything should work fine.

**Comment by DiMurer** [comment]
All fixes Done.

File: ads/adfox.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+import {loadScript, validateData} from '../3p/3p';
+import {yandex} from './yandex';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function adfox(global, data) {
```
Comment (zhouyx) [inline ads/adfox.js]: `adfox()` is not been imported elsewhere. no need to `export` here.

File: ads/yandex.md [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+```html
+<amp-ad width="300" height="250"
+    type="yandex"
+    data-block-id="R-A-123456-1">
+</amp-ad>
+```
+
+## Configuration
+
+For semantics of configuration, please see ad network documentation.
```
Comment (zhouyx) [inline ads/yandex.md]: please link the network documentation from here.

File: examples/ads.amp.html (hunk 19 above)
Comment (zhouyx) [inline examples/ads.amp.html]: works now. Thanks

File: ads/adfox.js (hunk 20 above)
Comment (DiMurer) [inline ads/adfox.js]: Are you sure about this?
If I remove the text, I get this error:
image

**Comment by DiMurer** [comment]
Done

File: ads/adfox.js (hunk 20 above)
Comment (zhouyx) [inline ads/adfox.js]: Oops. I was definitely wrong on this : ) Adding two ad networks together is so tricky : P

**Comment by zhouyx** [comment]
@DiMurer LGTM. Thanks for the Pull Request. Merged

**Pull request merged**